### DISCUSSION TITLE
Creating a second db instance in a separate isolate. Touches different tables. Safe?

### DISCUSSION THREAD

**Discussion by willhaslett** [simolus3/drift#1271] [Q&A]
I am writing to the db periodically whether the app is running or not. To achieve this, I'm using [alarm_manager](https://pub.dev/packages/android_alarm_manager), which gets things done in the background using a separate isolate from the app per se. In the app, I'm using a `Provider` (via Riverpod) to make the db singleton available throughout the widget tree. That instance is not available to the background isolate.

Here's the question: am I safe in spawning a temporary instance of the db (temporary as in, create it with `var` then set it to null when done), or am I exposing myself to the possibility of db corruption? The warning from the plugin mentions possible race conditions/corruptions if I'm using the same `QueryExecutor`. Is that the case with my setup?

Even if it is the case that my setup exposes me to this risk, am I in the clear if the app never touches the tables that the background process does?

All insights appreciated!

**Comment by simolus3** [comment]
sqlite3 in general (and in the specific way in which is has been compiled for `sqlite3_flutter_libs`) supports accessing the same database across different threads without risking corruption.
The main problem is moor's migration implementation, which works like this:

1. Check the current user version from the database
2. Compare it with the current version and run a migration if necessary
3. Set the user version in the database to the current version

These steps don't run atomically because that may alter how foreign keys work IIRC. So if you open the same database at the same time, there's a slight chance that the migrations end up interfering with each other. This is the only problem that you need to be aware of here.

So in the end, there's no risk if no two migrations are running at the same time. If you know that your main isolate runs and uses the database after your app gets installed or upgraded for instance, there's nothing to worry about. But if the two isolates may be opening _and_ upgrading the database at the same time, I'm afraid you probably need manual synchronization logic to avoid issues.

That said, most of theses issues came up before moor had support for isolates, and they were pretty much always caused by users accidentally opening the same database _very_ frequently on the same isolate. That's the reasons for the strong warnings in the docs, I've never actually seen this problem caused by two isolates.

**Reply by willhaslett** [reply]
Thank you, Simon! That clears it right up. It has been working great, just wanted to make sure that I wasn't headed for that 1 in 100,000 race condition blow up!
